numba/core/compiler.py

```diff
@@ -646,49 +652,35 @@ def compile_ir(typingctx, targetctx, func_ir, args, return_type, flags,
         # cannot compile it. As a result the following approach is taken:
         # 1. Create some new flags that copy the original ones but switch
         #    off rewrites.
-        # 2. Compile with 1. to get a compile result
-        # 3. Try and compile another compile result but this time with the
-        #    original flags (and IR being rewritten).
-        # 4. If 3 was successful, use the result, else use 2.
+        # 2. If rewrites are on in original flags try to compile with a copy of
+        #    func_ir and original flags (with possibly IR being rewritten)
+        #    ignoring compile exceptions
+        # 3. If above succeeded and we have compile result return it
+        # 4. Othwerise try and compile another compile result with rewrites off
+        #    and propagating exceptions up the stack.
 
-        # create flags with no rewrites
         norw_flags = copy.deepcopy(flags)
         norw_flags.no_rewrites = True
 
-        def compile_local(the_ir, the_flags):
-            pipeline = pipeline_class(typingctx, targetctx, library,
-                                      args, return_type, the_flags, locals)
-            return pipeline.compile_ir(func_ir=the_ir, lifted=lifted,
-                                       lifted_from=lifted_from)
-
-        # compile with rewrites off, IR shouldn't be mutated irreparably
-        norw_cres = compile_local(func_ir.copy(), norw_flags)
-
-        # try and compile with rewrites on if no_rewrites was not set in the
-        # original flags, IR might get broken but we've got a CompileResult
-        # that's usable from above.
         rw_cres = None
         if not flags.no_rewrites:
-            # Suppress warnings in compilation retry
-            with warnings.catch_warnings():
-                warnings.simplefilter("ignore", errors.NumbaWarning)
-                try:
-                    rw_cres = compile_local(func_ir.copy(), flags)
-                except Exception:
-                    pass
-        # if the rewrite variant of compilation worked, use it, else use
-        # the norewrites backup
+            try:
+                rw_cres = compile_local(func_ir.copy(), flags)
+            except Exception:
+                pass
+
         if rw_cres is not None:
-            cres = rw_cres
-        else:
-            cres = norw_cres
-        return cres
+            return rw_cres
+
+        # Suppress warnings in compilation retry
+        with warnings.catch_warnings():
+            warnings.simplefilter("ignore", errors.NumbaWarning)
+            norw_cres = compile_local(func_ir.copy(), norw_flags)
+
```

Comment:
I think this logic is now backwards compared to the original? The original compiled a version without rewrites and then subsequently compiled one with rewrites https://github.com/numba/numba/blob/master/numba/core/compiler.py#L664-L685 this was because the rewrites could mutate the IR irreparably (even if `deepcopy`'d).

Reply:
> If the columns dtype dict is a constant would partial application work so as to keep most of the pandas interface but avoid having the `build_map` in the bytecode? Also slightly puzzled as to how this is adding much compilation time...

@stuartarchibald Yes, exactly removing the `build_map` from the code resolves it. Answering your question, it probably adds to compile time because the IR size itself grows in quadratic fashion of the dict size, e.g. following times were observed for DF with 64 columns:

id | pipeline | pass_name | target_func | time
-- | -- | -- | -- | --
50 | nopython | nopython_type_inference | csv_reader_py$4 | 18.9416706
895 | object | object_mode_back_end | csv_reader_py$3 | 13.9494324
885 | object | object_mode_back_end | csv_reader_py$3 | 10.2176549
39 | object | object_mode_back_end | csv_reader_py$3 | 9.8354137
49 | object | object_mode_back_end | csv_reader_py$3 | 9.0511516

And tracing pylowering itself confirmed it makes most of `object_mode_back_end` pass times for csv_reader_py$3:
> 	Line 82: _py_lowering_stage: func=csv_reader_py$3 took: 9.83552408
> 	Line 104: _py_lowering_stage: func=csv_reader_py$3 took: 9.05098248
> 	Line 1814: _py_lowering_stage: func=csv_reader_py$3 took: 10.21686435
> 	Line 1836: _py_lowering_stage: func=csv_reader_py$3 took: 13.94962978

So if we look at llvm IR of the csv_reader_py$3 we could find there blocks similar to these:
```
### the following block checks if the variable was loaded and is not null
B12.endif.endif.endif.end...endif.8.endif:
  %".1071" = load i8*, i8** %"$40build_const_key_map.15"
  %".1072" = icmp eq i8* null, %".1071"                             
  br i1 %".1072", label %"B12.endif.endif.endif.end...endif.8.endif.if", label %"B12.endif.endif.endif.end...endif.8.endif.endif", !prof !0
  
### the block below is executed at failure (when in the above %".1071" is null)
B12.endif.endif.endif.end...endif.8.endif.if:       
  call void @"PyErr_SetString"(i8* @"PyExc_NameError", i8* bitcast ([48 x i8]* @".const.name '$40build_const_key_map.15' is not defined" to i8*))
  %".1075" = load i8*, i8** %"names"
  call void @"Py_DecRef"(i8* %".1075")
  %".1077" = load i8*, i8** %"$22build_list.6"
  call void @"Py_DecRef"(i8* %".1077")
  %".1079" = load i8*, i8** %"$30load_attr.10"
  call void @"Py_DecRef"(i8* %".1079")
  %".1081" = load i8*, i8** %"$40build_const_key_map.15"    ### since we raise exception, we call 
  call void @"Py_DecRef"(i8* %".1081")            ### PyLower::cleanup_vars() and decref all live variables
  %".1083" = load i8*, i8** %"skiprows"
  call void @"Py_DecRef"(i8* %".1083")
...
  %".1097" = load i8*, i8** %"$40build_const_key_map.18"    
  call void @"Py_DecRef"(i8* %".1097")
...
  %".1103" = load i8*, i8** %"$40build_const_key_map.16"
  call void @"Py_DecRef"(i8* %".1103")
  %".1105" = load i8*, i8** %"$26load_attr.8"
  call void @"Py_DecRef"(i8* %".1105")
  ret i32 -1
```

Reply:
> I think this logic is now backwards compared to the original? The original compiled a version without rewrites and then subsequently compiled one with rewrites https://github.com/numba/numba/blob/master/numba/core/compiler.py#L664-L685 this was because the rewrites could mutate the IR irreparably (even if `deepcopy`'d).

@kozlov-alexey was wondering if you had a view on the above? Thanks.

Reply:
@stuartarchibald Yes, I think we should skip the changes in `compile_local` for now, since it requires immutable IR anyway.  So if it's OK with you we can stick to the PR name and just fix issue locally in objmode dispatcher. That will remove objmode recompilations keeping the rest of compile logic the same.